File: go/2015/01/solution.go
// Package solution contains solutions to part 1 and 2
package solution

// Part1 returns the solution for day 1 part 1
func Part1(input string) int {
	floor := 0
	for _, v := range input {
		switch v {
		case '(':
			floor++
		case ')':
			floor--
		}
	}
	return floor
}

// Part2 returns the solution for day 1 part 2
func Part2(input string) int {
	floor := 0
	for i, v := range input {
		switch v {
		case '(':
			floor++
		case ')':
			floor--
		}
		if floor <= -1 {
			return i + 1
		}
	}
	return 0
}

File: python/2015/01/solution.py
def Part1(input: str) -> int:
    floor = 0
    for c in input:
        match c:
            case "(":
                floor += 1
            case ")":
                floor -= 1
    return floor


def Part2(input: str) -> int:
    floor = 0
    index = 0
    for c in input:
        index += 1
        match c:
            case "(":
                floor += 1
            case ")":
                floor -= 1
        if floor < 0:
            break
    return index
